core/src/test/java/org/apache/calcite/test/JdbcTest.java

```diff
@@ -461,6 +464,43 @@ private static String q(String s) {
     connection.close();
   }
 
+  /**
+   * <p>Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-3423">[CALCITE-3423]
+   * Support using CAST operation and BOOLEAN type value in table macro</a>. */
+  @Test public void testTableMacroWithCastOrBoolean() throws SQLException {
+    Connection connection =
+        DriverManager.getConnection("jdbc:calcite:");
+    addTableMacro(connection, Smalls.STR_METHOD);
+    // check for cast
+    ResultSet resultSet = connection.createStatement().executeQuery(
+        "select * from table(\"s\".\"str\"(MAP['a', 1, 'baz', 2], cast(1 as bigint))) as t(n)");
+    assertThat(CalciteAssert.toString(resultSet),
+        equalTo("N={'a'=1, 'baz'=2}\n"
+            + "N=1               \n"));
+    // check for Boolean type
+    resultSet = connection.createStatement().executeQuery(
+        "select * from table(\"s\".\"str\"(MAP['a', 1, 'baz', 2], true)) as t(n)");
+    assertThat(CalciteAssert.toString(resultSet),
+        equalTo("N={'a'=1, 'baz'=2}\n"
+            + "N=true            \n"));
+    // check for nested cast
+    resultSet = connection.createStatement().executeQuery(
+        "select * from table(\"s\".\"str\"(MAP['a', 1, 'baz', 2],"
+            + "cast(cast(1 as int) as varchar(1)))) as t(n)");
+    assertThat(CalciteAssert.toString(resultSet),
+        equalTo("N={'a'=1, 'baz'=2}\n"
+            + "N=1               \n"));
+
+    resultSet = connection.createStatement().executeQuery(
```

Comment:
For this case see the discussion in https://issues.apache.org/jira/browse/CALCITE-3423

Reply:
Can we also add a test case for implicit type coercion ?

Reply:
I tried, but failed to add an implicit type coercion case.
It seems that `TableMacro` is invoked at "compile time" , mis-match type will cause `SQLException`.
For example, if we add the code snippet below
```
// check for implicit type coercion
    addTableMacro(connection, Smalls.VIEW_METHOD);
    resultSet = connection.createStatement().executeQuery(
        "select * from table(\"s\".\"view\"(5))) as t(n)");
    assertThat(CalciteAssert.toString(resultSet),
        equalTo("N=1\n"
            + "N=3\n"
            + "N=5\n"));
```
we'll get exception when running the case
```
java.sql.SQLException: Error while executing SQL "select * from table("s"."view"(5))) as t(n)": parse failed: Encountered ")" at line 1, column 35.
```
```
at org.apache.calcite.prepare.CalcitePrepareImpl.prepare2_(CalcitePrepareImpl.java:617)
	at org.apache.calcite.prepare.CalcitePrepareImpl.prepare_(CalcitePrepareImpl.java:502)
	at org.apache.calcite.prepare.CalcitePrepareImpl.prepareSql(CalcitePrepareImpl.java:472)
	at org.apache.calcite.jdbc.CalciteConnectionImpl.parseQuery(CalciteConnectionImpl.java:231)
	at org.apache.calcite.jdbc.CalciteMetaImpl.prepareAndExecute(CalciteMetaImpl.java:550)
	at org.apache.calcite.avatica.AvaticaConnection.prepareAndExecuteInternal(AvaticaConnection.java:675)
	at org.apache.calcite.avatica.AvaticaStatement.executeInternal(AvaticaStatement.java:156)
```

And the implicit type coercion case is already exist in `TableFunctionTest.testUserDefinedTableFunction4`

Reply:
Thanks @yanlin-Lynn , the code basically looks good, can you also add support of implicit type coercion for user defined table macro ? Because this patch add a explicit cast support for table macro, i think we can remove the logic of for user defined table macro in `TypeCoercionImpl#userDefinedFunctionCoercion`:
```java
// User defined table macro only allows literals.
    // we should support this in the future.
    if (function instanceof SqlUserDefinedTableMacro) {
      return false;
    }
```

Reply:
Sure, I'll update the PR soon.